src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,393 @@
+    in_data = OPENSSL_malloc(sizeof(TPM2B_MAX_BUFFER));
+    if (tpm2Data == NULL) {
+        ERR(tpm2_do_cipher, ERR_R_MALLOC_FAILURE);
+        goto error;
+    }
+    memcpy(in_data->buffer, in, inl);
+    in_data->size = inl;
+
+    /* Get mode value */
+    mode = tpm2_get_cipher_mode(ctx, tpm2Data);
+
+    /* Get IV and Enc*/
+    iv_in.size = EVP_CIPHER_CTX_iv_length(ctx);
+#if OPENSSL_VERSION_NUMBER < 0x10100000
+    memcpy(iv_in.buffer, ctx->iv, iv_in.size);
+    // Note: Openssl (encrypt:1) != TSS (encrypt:0)
+    enc = !(ctx->encrypt);
+#else
+    memcpy(iv_in.buffer, EVP_CIPHER_CTX_iv(ctx), iv_in.size);
+    enc = !EVP_CIPHER_CTX_encrypting(ctx);
+#endif
+
+    /* Trying to encrypt */
+    ret = Esys_EncryptDecrypt2( eactx.ectx,
+                                keyHandle,
+                                ESYS_TR_PASSWORD,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                in_data,
+                                enc,
+                                mode,
+                                &iv_in,
+                                &out_data,
+                                &iv_out );
+    if (ret == TPM2_RC_COMMAND_CODE) {
+        DBG("Esys_EncryptDecrypt2 : FAILED\n");
+        ret = Esys_EncryptDecrypt( eactx.ectx,
+                                   keyHandle,
+                                   ESYS_TR_PASSWORD,
+                                   ESYS_TR_NONE,
+                                   ESYS_TR_NONE,
+                                   enc,
+                                   mode,
+                                   &iv_in,
+                                   in_data,
+                                   &out_data,
+                                   &iv_out );
+        if(ret == TPM2_RC_SUCCESS) {
+            DBG("Esys_EncryptDecrypt  : SUCCESS\n");
+        }
+        else {
+            DBG("Esys_EncryptDecrypt  : FAILED\n");
+        }
+    }
+    ERRchktss(tpm2_do_cipher, ret, goto error);
+
+    /* Copy out_data : TPM2B_MAX_BUFFER to unsigned char* */
+    memcpy(out, out_data->buffer, out_data->size);
+    out[out_data->size] = '\0';
+
+    /* Close TPM session */
+    if (keyHandle != ESYS_TR_NONE) {
+        if (tpm2Data->privatetype == KEY_TYPE_HANDLE) {
+            Esys_TR_Close(eactx.ectx, &keyHandle);
+        } else {
+            Esys_FlushContext(eactx.ectx, keyHandle);
+        }
+    }
+    esys_auxctx_free(&eactx);
+
+    return out_data->size;
+
+error :
+    /* Close TPM session */
+    if (keyHandle != ESYS_TR_NONE) {
+        if (tpm2Data->privatetype == KEY_TYPE_HANDLE) {
+            Esys_TR_Close(eactx.ectx, &keyHandle);
+        } else {
+            Esys_FlushContext(eactx.ectx, keyHandle);
+        }
+    }
+    esys_auxctx_free(&eactx);
+
+    if (tpm2Data)
+        OPENSSL_free(tpm2Data);
+    return 0;
+}
+
+static int
+tpm2_cipher_cleanup(EVP_CIPHER_CTX *ctx)
```

Comment:
Since this is basically a no-op, can you just not declare this and pass NULL in the struct ?